### PR TITLE
Update string manipulation and comparison tests && add R-4 tests

### PR DESCRIPTION
**Pull request by Rende11**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -22,9 +22,9 @@
   },
   "scripts": {
     "generateParser": "cd src/parser; rm -rf ./generated/*; java -Xmx500M -cp \"../../antlr-4.7.1-complete.jar:$CLASSPATH\" org.antlr.v4.Tool -o generated -Dlanguage=JavaScript FHIRPath.g4",
-    "test:unit": "jest && npm run lint",
+    "test:unit": "jest",
     "test:demo": "npm run build && cd demo && npm run build && grunt test:e2e",
-    "test": "npm run test:unit && npm run test:demo",
+    "test": "npm run test:unit && npm run test:demo npm run lint",
```
[the remark this answers is not in the document]
Comment (Rende11) [reply package.json]: fixed

**Comment by plynchnlm** [comment]
It looks like the r4 test structure has been flattened.  In the XML file there are "groups" of tests.   I think it would be helpful to add the concept of "groups" to our test runner, so that:

1) tests within a group would be nested, visually associating them with the group label
2) an entire group of tests could be disabled with one setting of "disabled: true".

File: test/fhirpath.test.js [hunk 2]
```diff
@@ -31,32 +32,47 @@ for (var i=0; i<items.length; i++) {
...
+              } else {
+                it(((t.desc || '') + ': ' + (e || '')) , () => {
+                  if (t.disableConsoleLog)
+                    console.log = function() {};
+                  if (!t.error && t.expression) {
+                    let res;
+                    if (_.has(t, 'inputfile')) {
+                      const filePath = __dirname + /resources/ + t.inputfile;
+                      if (fs.existsSync(filePath)) {
+                        const subjFromFile = JSON.parse(fs.readFileSync(filePath));
```
Comment (plynchnlm) [inline test/fhirpath.test.js]: There isn't a need to read these files in each time.  It should speed up the tests a little if you add a cache of the parsed file objects (with the filepath as the key).

File: test/fhirpath.test.js [hunk 3]
```diff
@@ -31,32 +32,47 @@ for (var i=0; i<items.length; i++) {
...
-                if (!t.error && t.expression) {
-                  var res = subj.evaluate(testcase.subject, e);
-                  expect(res).toEqual(t.result);
-                }
-                else {
-                  let exceptn = null;
-                  try {
-                    subj.evaluate(testcase.subject, e);
+            exprs.forEach((e)=> {
+              if (t.disable && t.disable === true) {
```
[the remark this answers is not in the document]
Comment (Rende11) [reply test/fhirpath.test.js]: fixed

File: test/fhirpath.test.js [hunk 4]
```diff
@@ -31,32 +32,47 @@ for (var i=0; i<items.length; i++) {
...
+                    }
+                    expect(e).not.toBe(null);
                   }
-                  expect(e).not.toBe(null);
-                }
-                if (t.disableConsoleLog)
-                  console.log = console.log;
-              });
+                  if (t.disableConsoleLog)
+                    console.log = console.log;
```
[the remark this answers is not in the document]
Comment (Rende11) [reply test/fhirpath.test.js]: fixed

File: test/fhirpath.test.js (hunk 2 above)
Comment (Rende11) [reply test/fhirpath.test.js]: I add object with data from resources files

File: test/fhirpath.test.js [hunk 5]
```diff
@@ -6,86 +6,90 @@ const _    = require('lodash');
...
+    if(_.has(resources, t.inputfile)) {
+      return  subj.evaluate(resources[t.inputfile], e);
+    } else {
+      const filePath = __dirname + /resources/ + t.inputfile;
+      if (fs.existsSync(filePath)) {
+        const subjFromFile = JSON.parse(fs.readFileSync(filePath));
+        resources[t.inputfile] = subjFromFile;
+        return subj.evaluate(subjFromFile, e);
+      } else {
+        throw new Error('Resource file isnt exists');
```
[the remark this answers is not in the document]
Comment (Rende11) [reply test/fhirpath.test.js]: fixed!
Thx =)

File: test/fhirpath.test.js [hunk 6]
```diff
@@ -6,86 +6,90 @@ const _    = require('lodash');
...
-    const testcase = yaml.safeLoad(fs.readFileSync( __dirname + '/cases/' + fileName, 'utf8'));
 
-    if((focus && focusFile.test(fileName)) || focus === false) {
-      let focusedTest = false;
-      for (let i=0, len=testcase.tests.length; i<len && !focusedTest; ++i) {
-        focusedTest = testcase.tests[i].focus;
+const calcExpression = (e, t, testcase) => {
+  if (_.has(t, 'inputfile')) {
+    if(_.has(resources, t.inputfile)) {
+      return  subj.evaluate(resources[t.inputfile], e);
```
[the remark this answers is not in the document]
Comment (Rende11) [reply test/fhirpath.test.js]: replaced to "fhirpath"

File: test/fhirpath.test.js [hunk 7]
```diff
@@ -6,86 +6,90 @@ const _    = require('lodash');
...
 
-  // Only process yaml files
-  if (endWith(fileName, '.yaml')) {
-    const testcase = yaml.safeLoad(fs.readFileSync( __dirname + '/cases/' + fileName, 'utf8'));
 
-    if((focus && focusFile.test(fileName)) || focus === false) {
-      let focusedTest = false;
-      for (let i=0, len=testcase.tests.length; i<len && !focusedTest; ++i) {
-        focusedTest = testcase.tests[i].focus;
+const calcExpression = (e, t, testcase) => {
```
[the remark this answers is not in the document]
Comment (Rende11) [reply test/fhirpath.test.js]: I replace all "e, t" to expression and test.

**Comment by plynchnlm** [comment]
On test is still failling:
```
6.4_collection.yaml › ** in operand is coll: c in d

    expect(received).not.toBe(expected) // Object.is equality

    Expected: null
    Received: null

      65 |           exception = error;
      66 |         }
    > 67 |         expect(exception).not.toBe(null);
         |                               ^
      68 |       }
      69 |       if (test.disableConsoleLog)
      70 |         console.log = console_log;

      at Object.toBe (test/fhirpath.test.js:67:31)
```

**Comment by Rende11** [comment]
Now build is "green" )

**Pull request merged**